### ISSUE TITLE
python-setuptools package for RHEL 7 should provide python2-setuptools

### ISSUE THREAD

**Issue by mtorromeo** [saltstack/salt#56212]
When upgrading to salt 3000 on CentOS 7 I encountered incompatibility issues with some packages in EPEL that require python2-setuptools.

For example the certbot package depends on python2-josepy which in turn depends on python2-setuptools. Since [CentOS' own package](https://centos.pkgs.org/7/centos-x86_64/python-setuptools-0.9.8-7.el7.noarch.rpm.html) provides python2-setuptools, this is totally fine but the package provided by salt's repo doesn't.

I temporarily solved this by building my [own fixed package](https://github.com/artera/rpm-python-setuptools) which just adds the missing `Provides`.

**Comment by dmurphy18** [comment]
@mtorromeo The issue is support for earlier versions of Redhat 7, for example: Redhat 7.2 where python2-setuptools did not originally exist, and still supporting that version. If I recall correctly originally did have python2-setuptools and had to revert to python-setuptools for packages from early versions of Redhat 7.  Shall review minimum versions of Redhat 7 support to see if we can address this, but Redhat altering their original minimum versions ISO's for Python 3 support which did not exist in them originally, may make this issue difficult.

**Comment by mtorromeo** [comment]
Yeah, I noticed that even with my altered package I still have to manually remove the previous version of python-setuptools to resolve the conflict but at least I don't end up in a situation with a permanent unresolved dependency. I'm not sure how to fix this cleanly.

**Comment by dmurphy18** [comment]
The problem is that rpm name resolution depends on the name of the rpm rather than what it provides.  Hoping that when I review when python2-setuptools was introduced that it is on an old enough version that SaltStack can pick that as the minimum supported version.  Will review that next week since going PTO rest of this week.

**Comment by waynew** [comment]
@dmurphy18 were you able to confirm anything here?

**Comment by waynew** [comment]
Had a conversation with @dmurphy18 - from what I understand, this will be resolved by packaging salt with salt-bin, which we'll be doing in Sodium, so there shouldn't be any issues here. 

Based on that, I'm going to go ahead and close this issue.

**Comment by onmeac** [comment]
Please re-open and fix :), adding `Provides` does work, but a typo was made in its value.

This issue is breaking salt upgrades to (in our environment) version 3000, but only on minions which also have certbot installed.


```
# https://github.com/artera/rpm-python-setuptools/blob/master/python-setuptools.spec
- Provides:       python2-setuptoos = %{version}-%{release}
+ Provides:       python2-setuptools = %{version}-%{release}
```



### with typo
```
Error: Package: python2-josepy-1.2.0-1.el7.noarch (epel)
           Requires: python2-setuptools
           Available: python-setuptools-0.9.8-7.el7.noarch (repo_base)
               python2-setuptools = 0.9.8-7.el7
           Installed: python-setuptools-36.6.0-3.ius.el7.noarch (@/python-setuptools-36.6.0-3.ius.el7.noarch)
              ~python-setuptools = 36.6.0-3.ius.el7
           Available: python-setuptools-36.6.0-2.ius.el7.noarch (repo_saltstack)
              ~python-setuptools = 36.6.0-2.ius.el7
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```



### without typo
```
 Dependencies Resolved

 ==================================================================================================================================================================================
  Package                                              Arch                               Version                                      Repository                             Size
 ==================================================================================================================================================================================
 Installing:
  certbot                                              noarch                             1.0.0-1.el7                                  epel                                   41 k
 Installing for dependencies:
  pyOpenSSL                                            x86_64                             0.13.1-4.el7                                 repo_base                             135 k
  python-cffi                                          x86_64                             1.6.0-5.el7                                  repo_base                             218 k
  python-enum34                                        noarch                             1.0.4-1.el7                                  repo_base                              52 k
  python-idna                                          noarch                             2.4-1.el7                                    repo_base                              94 k
  python-ndg_httpsclient                               noarch                             0.3.2-1.el7                                  epel                                   43 k
  python-ply                                           noarch                             3.4-11.el7                                   repo_base                             123 k
  python-pycparser                                     noarch                             2.14-1.el7                                   repo_base                             104 k
  python-requests-toolbelt                             noarch                             0.8.0-1.el7                                  epel                                   77 k
  python-zope-component                                noarch                             1:4.1.0-5.el7                                epel                                  228 k
  python-zope-event                                    noarch                             4.0.3-2.el7                                  epel                                   79 k
  python-zope-interface                                x86_64                             4.0.5-4.el7                                  repo_base                             138 k
  python2-acme                                         noarch                             1.0.0-1.el7                                  epel                                   74 k
  python2-certbot                                      noarch                             1.0.0-1.el7                                  epel                                  358 k
  python2-configargparse                               noarch                             0.11.0-1.el7                                 epel                                   30 k
  python2-cryptography                                 x86_64                             1.7.2-2.el7                                  repo_base                             502 k
  python2-distro                                       noarch                             1.2.0-3.el7                                  epel                                   29 k
  python2-future                                       noarch                             0.18.2-2.el7                                 epel                                  806 k
  python2-josepy                                       noarch                             1.2.0-1.el7                                  epel                                   89 k
  python2-mock                                         noarch                             1.0.1-10.el7                                 epel                                   92 k
  python2-parsedatetime                                noarch                             2.4-5.el7                                    epel                                   78 k
  python2-pyasn1                                       noarch                             0.1.9-7.el7                                  repo_base                             100 k
  python2-pyrfc3339                                    noarch                             1.0-2.el7                                    epel                                   13 k
  python2-six                                          noarch                             1.9.0-0.el7                                  epel                                  2.9 k
  pytz                                                 noarch                             2016.10-2.el7                                repo_base                              46 k

 Transaction Summary
 ==================================================================================================================================================================================
 Install  1 Package (+24 Dependent packages)

 Total download size: 3.5 M
 Installed size: 16 M
 Is this ok [y/d/N]: y


```

**Comment by dmurphy18** [comment]
@onmeac That typo is in IUS code and revision 3, the code used by SaltStack is revision 2 which does not have the statement and hence does not have the typo.

However, will add the statement for Provides with the typo corrected and it should appear in the next point release which is immanent, this one is just slipping under the wire.

**Comment by waynew** [comment]
@dmurphy18 is there anything more needed here? If not, please go ahead and close this again.

**Comment by dmurphy18** [comment]
@onmeac This was fixed in the latest Salt 3000.1 Python2 Redhat 7 release, a section for the spec file
```
 18 BuildArch:      noarch
 19 BuildRequires:  %{python}-devel
 20 Requires:       %{python}
 21 Provides:       python2-setuptools = %{version}-%{release}
 22 Provides:       python-setuptools = %{version}-%{release}
 23 # Keep the python-distribute name active for a few releases.  Eventually we'll
 24 # want to get rid of the Provides and just keep the Obsoletes
 25 Provides:       %{python}-distribute = %{version}-%{release}
 26 Obsoletes:      %{python}-distribute <= 0.6.49-2.ius%{?dist}
 27 %if 0%{?with_check}
 28 # we don't have IUS versions of these yet
 29 BuildRequires:  %{python}-pytest
 30 BuildRequires:  %{python}-mock
 31 %endif 
```

Was going to wait for salt-bin to address this, but the opportunity came for me to slip the fix in, and hopefully this will resolve your issue.

Hence closing this issue.
